### PR TITLE
move handshake proxy to nginx

### PR DESCRIPTION
**Pull request by kwypchlo**
Move handshake proxy endpoint to nginx with help of lua scripts. This allows us to keep it as low level as possible, removing any potential overhead from running requests through node server.

### CODE REVIEW

File: docker/nginx/conf.d/client.conf [hunk 1]
```diff
@@ -127,10 +128,42 @@ server {
...
-		proxy_pass http://handshake-api:3100;
+		set $skylink ''; # placeholder for the raw 46 bit skylink
+		set $rest ''; # placeholder for the rest of the url that gets appended to skylink (path and args)
+		
+		# resolve handshake domain by requesting to /hnsres endpoint and assign correct values to $skylink and $rest
+		access_by_lua_block {
+			local json = require('cjson')
+			local hns_domain_name, request_uri_rest = string.match(ngx.var.request_uri, "/hns/([^/?]+)(.*)")
+			local hnsres_res = ngx.location.capture("/hnsres/" .. hns_domain_name)
+			local hnsres_json = json.decode(hnsres_res.body)
```
Comment (peterjanbrone) [inline docker/nginx/conf.d/client.conf:141]: You want to handle responses with status != 200 here differently. Lua is debug hell if you don't handle errors properly

File: docker/nginx/conf.d/client.conf [hunk 2]
```diff
@@ -127,10 +128,42 @@ server {
...
+
+			ngx.var.skylink = skylink
+			if request_uri_rest == "" or (request_uri_rest == "/" and skylink_rest ~= "") then
+				ngx.var.rest = skylink_rest
+			else
+				ngx.var.rest = request_uri_rest
+			end
+		}
+
+		# overwrite the Cache-Control header to only cache for 60s in case the domain gets updated
```
Comment (peterjanbrone) [inline docker/nginx/conf.d/client.conf:152]: What does this mean exactly after we've proxy passed it to localhost again which should handle caching right? Isn't the setting of the header redundant at this point?

**Review by peterjanbrone — APPROVED** [review]
I like it!

File: docker/nginx/conf.d/client.conf (hunk 2 above)
Comment (kwypchlo) [reply docker/nginx/conf.d/client.conf:152]: this is for browser caching - it only sets the header that is consumed by browser

the cache you are referring to is proxy cache inside of nginx and it's not affected by that line

File: docker/nginx/conf.d/client.conf (hunk 1 above)
Comment (kwypchlo) [reply docker/nginx/conf.d/client.conf:141]: yeah I should probably 404 here

**Pull request merged**